### PR TITLE
Add RmlUi submodule

### PR DESCRIPTION
**Pull request by Venuskan**
did it.

### CODE REVIEW

**Comment by jonesmz** [comment]
You did everything correctly, except that's not the correct commit of the rmlui submodule.

It should be this commit:
https://github.com/mikke89/RmlUi/commit/76245e8d615e5771ab6dd01ac4ea3c59038e6071

What you should do is
```bash
cd 3rdparty/rmlui
git checkout 4.3
cd ../../
git add 3rdparty/rmlui
git commit --amend --no-edit
git push -f
```
to fix the pull request.

The reason why i say that it is not the correct commit is because whenever possible, we prefer "official releases" of our dependencies. RmlUI officially released version 4.3, which is stored in git as a "tag". You can see the list of tags using `git tag -l`

**Comment by jonesmz** [comment]
While this is building properly on Linux (and the failures on Mac are expected for unrelated reasons).

We are seeing a new failure for Windows.

```
CMake Error at C:/Program Files/CMake/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Freetype (missing: FREETYPE_LIBRARY) (found version
  "2.10.0")
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  C:/Program Files/CMake/share/cmake-3.22/Modules/FindFreetype.cmake:162 (find_package_handle_standard_args)
  3rdparty/RmlUi/CMakeLists.txt:325 (find_package)


-- Configuring incomplete, errors occurred!
See also "D:/a/osp-magnum/osp-magnum/build/CMakeFiles/CMakeOutput.log".
See also "D:/a/osp-magnum/osp-magnum/build/CMakeFiles/CMakeError.log".
Error: Process completed with exit code 1.
```

Looks like RmlUI explictly depends on "freetype", and the rule for osp-magnum is that we don't want to require third party dependencies to need to be installed ahead of time to build the code.

So you'll need to add another git submodule for freetype to get rmlui to build.

**Comment by Venuskan** [comment]
like this?

File: 3rdparty/CMakeLists.txt [hunk 1]
```diff
@@ -122,6 +122,8 @@ ADD_SUBDIRECTORY(entt EXCLUDE_FROM_ALL)
 ADD_SUBDIRECTORY(toml11 EXCLUDE_FROM_ALL)
 ADD_SUBDIRECTORY(spdlog EXCLUDE_FROM_ALL)
 ADD_SUBDIRECTORY(longeronpp EXCLUDE_FROM_ALL)
+ADD_SUBDIRECTORY(RmlUi EXCLUDE_FROM_ALL)
+Add_SUBDIRECTORY(freetype EXCLUDE_FROM_ALL)
```
Comment (jonesmz) [inline 3rdparty/CMakeLists.txt:126]: Freetype should be before rmlui

**Comment by jonesmz** [comment]
Mostly.

Please use the freetype mirror on github, so that continuous integration doesn't have to use as much download bandwidth to check out the repository.

And set it to the git-tag "VER-2-11-1"

**Comment by Venuskan** [comment]
does the CI failure still occur?

**Comment by jonesmz** [comment]
(ADD_SUBDIRECTORY):
  ADD_SUBDIRECTORY given source "RmlUi" which is not an existing directory.

**Comment by jonesmz** [comment]
(ADD_SUBDIRECTORY):
  ADD_SUBDIRECTORY given source "RmlUi" which is not an existing directory.

**Comment by jonesmz** [comment]
(ADD_SUBDIRECTORY):
  ADD_SUBDIRECTORY given source "RmlUi" which is not an existing directory.

**Comment by jonesmz** [comment]
Please make sure you can build the code on your machine before trying to push the pull request

**Comment by jonesmz** [comment]
Please make sure you can build the code on your machine before trying to push the pull request

**Comment by Capital-Asterisk** [comment]
The RmlUi submodule and directory aren't present, though freetype is there. As it seems, modifying .gitmodules doesn't properly add or remove submodules. There may be some issues with trying re-adding a deleted submodule however.

I'm not sure if this is sketchy or not, but you can revert back to older commits using git reset. 1a46823 is the shortened hash of the "clean up rendering code" commit. reset --hard is a bit of a dangerous command as it deletes uncommitted changes.
```bash
git reset --hard 1a46823
```
It should now be good to try adding the RmlUi submodule.

Afterwards, use cherry-pick to re-add your CMake changes from commit (d04d1633) and freetype submodule (6d23590).
```bash
git cherry-pick 6d23590 d04d1633
```

When pushing changes to github, force push is required since this rewrites the history of your master branch.
```bash
git push -f origin master
```

**Pull request closed without merging**